### PR TITLE
Finalize First API Types

### PR DESCRIPTION
**Pull request by KtorZ**
# Issue Number

#53 

# Overview

- [ ] I have move validation logic and representation of various types back to the wallet primitives
- [ ] I have added extra tests to validate error branches for a few JSON serialization

# Comments

### CODE REVIEW

File: src/Cardano/Wallet/Api/Types.hs [hunk 1]
```diff
@@ -3,219 +3,153 @@
...
--- | A Wallet State representation in the API. It can be serialized to and from
--- JSON as follows:
+-- | Options for encoding a wallet state. It can be serialized to and from JSON
+-- as follows:
 --
 -- >>> Aeson.encode Ready
 -- {"status":"ready"}
 --
--- >>> Aeson.encode $ Restoring (Percentage 14)
+-- >>> Aeson.encode $ Restoring (MeasuredIn 14)
```
Comment (Anviking) [inline src/Cardano/Wallet/Api/Types.hs]: Auxiliary: On the value-level, the name is kind of confusing. Could we rename `MeasuredIn` to [`Quantity`](https://en.wikipedia.org/wiki/Physical_quantity)?

File: src/Data/MeasuredIn.hs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Representation of values with an associated (free) unit of measure. Useful to
+-- disambiguate primitive types like 'Int' or 'String' which can be in different
+-- bases depending on the context.
+
+module Data.MeasuredIn
```
Comment (Anviking) [inline src/Data/MeasuredIn.hs]: Think separating into its own module can make sense 👍

File: src/Data/MeasuredIn.hs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                                Percentage
+-------------------------------------------------------------------------------}
+
+deriving newtype instance Integral i => Ord (MeasuredIn "percent" i)
+
+instance Integral i => Bounded (MeasuredIn "percent" i) where
+    minBound = MeasuredIn 0
+    maxBound = MeasuredIn 100
+
+instance {-# OVERLAPS #-} (Integral i, FromJSON i) => FromJSON (MeasuredIn "percent" i) where
```
Comment (Anviking) [inline src/Data/MeasuredIn.hs]: `{-# OVERLAPS #-}` why is this needed?

Also seems weird to have both
1) Very high-tech & generlized `MeasuredIn`, and
2) Special case for percent

File: src/Cardano/Wallet/Api/Types.hs [hunk 4]
```diff
@@ -3,219 +3,153 @@
...
     , _addressPoolGap :: !(ApiT AddressPoolGap)
-    , _balance :: !WalletBalance
-    , _delegation :: !WalletDelegation
+    , _balance :: !(ApiT WalletBalance)
+    , _delegation :: !(ApiT (WalletDelegation (ApiT PoolId)))
     , _name :: !(ApiT WalletName)
-    , _passphrase :: !WalletPassphraseInfo
-    , _state :: !WalletState
+    , _passphrase :: !(ApiT WalletPassphraseInfo)
+    , _state :: !(ApiT WalletState)
```
Comment (Anviking) [inline src/Cardano/Wallet/Api/Types.hs:76]: 👍 Will be interesting to see how this plays out now when our types are much more polymorphic (and able to handle this) than in cardano-sl.

File: src/Cardano/Wallet.hs [hunk 5]
```diff
@@ -276,9 +280,9 @@ data WalletMetadata = WalletMetadata
     , passphraseInfo
         :: !WalletPassphraseInfo
     , status
```
Comment (Anviking) [inline src/Cardano/Wallet.hs:282]: Ah this is the term used in the API spec.

Maybe change the record-field too:
```suggestion
    , state
```

File: src/Cardano/Wallet.hs [hunk 6]
```diff
@@ -301,22 +305,23 @@ walletNameMinLength = 1
...
 
-newtype WalletTimestamp = WalletTimestamp Microsecond
-    deriving (Eq, Ord, Show)
-
-data WalletStatus
+data WalletState
     = Ready
-    | Restoring
-    deriving (Eq, Show)
+    | Restoring !(MeasuredIn "percent" Word)
```
Comment (Anviking) [inline src/Cardano/Wallet.hs]: Hm… You have just defined Aeson instances on a `Cardano.Wallet` type, haven't you? :)

Comment (Anviking) [reply src/Cardano/Wallet.hs]: Ok, not really, but I'm confused by this 🤔 

Why have `ApiT` at all?

File: test/unit/Data/MeasuredInSpec.hs [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            (Left @String @(MeasuredIn "bytes" Int) msg)
+
+    it "fail to percent from JSON when out of bounds" $ do
+        let msg = "Error in $: PercentageOutOfBoundsError"
+        Aeson.eitherDecode "{\"unit\":\"percent\",\"quantity\":-14}" `shouldBe`
+            (Left @String @(MeasuredIn "percent" Int) msg)
+
+    it "can't make percentage out of bound" $ checkCoverage $ \case
+        n | n > 10 -> cover 25 True "n > 100" $
+            -- NOTE taking 10 * n to get bigger Int generated by QuickCheck
```
Comment (Anviking) [inline test/unit/Data/MeasuredInSpec.hs]: 🤷‍♂️

File: test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 8]
```diff
@@ -63,13 +67,13 @@ spec = do
...
-            roundtripAndGolden $ Proxy @ WalletDelegation
-            roundtripAndGolden $ Proxy @ WalletId
+            roundtripAndGolden $ Proxy @ (ApiT WalletBalance)
+            roundtripAndGolden $ Proxy @ (ApiT (WalletDelegation (ApiT PoolId)))
+            roundtripAndGolden $ Proxy @ (ApiT WalletId)
             roundtripAndGolden $ Proxy @ (ApiT WalletName)
-            roundtripAndGolden $ Proxy @ WalletBalance
-            roundtripAndGolden $ Proxy @ WalletPassphraseInfo
-            roundtripAndGolden $ Proxy @ WalletState
+            roundtripAndGolden $ Proxy @ (ApiT WalletBalance)
```
Comment (Anviking) [inline test/unit/Cardano/Wallet/Api/TypesSpec.hs:74]:
```suggestion
```

I think I messed up, this is a duplicate

Don't have to fix here, but could

**Review by Anviking — COMMENTED** [review]
First look. Looks good.

Mainly two points of concerns / confusion related to `MeasuredIn "percent"`.

- https://github.com/input-output-hk/cardano-wallet/pull/107#discussion_r267920703
- https://github.com/input-output-hk/cardano-wallet/pull/107#discussion_r267943565

File: src/Cardano/Wallet.hs (hunk 6 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet.hs]: > Why have `ApiT` at all?

Having `ApiT` allows us to avoid orphan instances, and to keep the following concerns separate:
1. How we define a type that we use internally.
2. How we represent values of that type on the wire.

For types that are only ever seen in the API, we can define `FromJSON` and `ToJSON` instances directly for those types, and the job is done.

For types that appear in the API, but are actually defined *elsewhere* (like `AddressPoolGap`), we still want to define `FromJSON` and `ToJSON` instances in the same way as the other types, but we don't want to place those instances in `Primitive` (to keep the separation of concerns).

Therefore, we need to wrap such types in a `newtype` wrapper, in order to be able to give them non-orphan instances. We could, of course, use a different `newtype` for each and every such type. But in the end, most of these wrappers would be the same. Hence we define `ApiT` as a common wrapper, and give each `ApiT a` a unique instance of `FromJSON` and `ToJSON`.

Comment (jonathanknowles) [inline src/Cardano/Wallet.hs]:
```hs
    | Restoring !(MeasuredIn "percent" Word)
```
I still think this seems very odd. Of course, the _JSON encoding_ of a percentage is an _object_ with a `quantity` and a `unit` of `percent`, but that's just the wire format.

I understand the general case where need to have a type that is polymorphic in some unit. For example, monetary amounts can be measured in different currency units, and so encoding these units in the type system makes sense, because the compiler will stop us from erroneously mixing amounts with different currency units.

But for a _percentage_, the units can never change, they'll always be "percent". Rather than write `MeasuredIn "percent" Word`, why not just have a simpler `Percentage` type that is unit-less, and then when we want to encode it as JSON, we can add the `"unit" : "percent"` property back in again?

It seems here that we're allowing the wire format to influence our choice of how to structure what could be a perfectly ordinary Haskell type.

File: src/Cardano/Wallet.hs [hunk 9]
```diff
@@ -301,22 +305,23 @@ walletNameMinLength = 1
...
-    = Delegated
-    | NotDelegated
-    deriving (Eq, Show)
+data WalletDelegation poolId
+    = NotDelegating
+    | Delegating !poolId
+    deriving (Generic, Eq, Show)
+
+newtype PoolId = PoolId
+    { getPoolId :: Text }
```
Comment (jonathanknowles) [inline src/Cardano/Wallet.hs:322]: Should this be a `UUID`?

File: src/Cardano/Wallet/Api/Types.hs [hunk 10]
```diff
@@ -3,219 +3,153 @@
...
+    } deriving (Eq, Generic, Show)
+
+data WalletBalance = WalletBalance
+    { _available :: !(MeasuredIn "lovelace" Natural)
+    , _total :: !(MeasuredIn "lovelace" Natural)
```
Comment (jonathanknowles) [inline src/Cardano/Wallet/Api/Types.hs]: What if we were to define a type synonym here:

```hs
type Amount x = MeasuredIn x Natural
```
Then we could write:
```hs
data WalletBalance = WalletBalance
    { _available :: !(Amount "lovelace")
    , _total :: !(Amount "lovelace")
```

File: src/Cardano/Wallet/Api/Types.hs (hunk 1 above)
Comment (jonathanknowles) [reply src/Cardano/Wallet/Api/Types.hs]: I would propose just defining a `Percentage` data type for this purpose. Percentages can only ever have a unit of `percent`. There's really no need for polymorphism. The question of how to encode a percentage in JSON in our API should not restrict how we define our internal Haskell type, in my opinion.

Alternatively, if we want to keep the units of measure scheme even for percentages, why not define the following function:

```hs
withUnit :: forall u a . a -> MeasuredIn (u :: Symbol) a
withUnit = MeasuredIn
```
We can then write:
```hs
Aeson.encode $ Restoring (withUnit @"percent" 14)
```

File: src/Data/MeasuredIn.hs [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- Representation of values with an associated (free) unit of measure. Useful to
+-- disambiguate primitive types like 'Int' or 'String' which can be in different
+-- bases depending on the context.
+
+module Data.MeasuredIn
+    ( -- * Polymorphic MeasuredIn
+      MeasuredIn(..)
```
Comment (jonathanknowles) [inline src/Data/MeasuredIn.hs]: Perhaps this is just nitpicking, but some of this styling seems inconsistent.

Some brackets have spaces before them, but others don't.

Given that we already have a well-defined style for imports (== spaces before parentheses), perhaps we can copy that style for exports? (Can make a proposal if you like. )

File: src/Data/MeasuredIn.hs [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+
+-- | Represents a value that has an associated unit of measure, based on some
+--   underlying type.
+--
+-- >>> newtype Amount = Amount (MeasuredIn "lovelace" Word32)
+newtype MeasuredIn (u :: Symbol) a = MeasuredIn a
+    deriving (Generic, Show, Eq)
+
+-- | Encode to JSON delegating the
```
Comment (jonathanknowles) [inline src/Data/MeasuredIn.hs]: Seems like an incomplete comment? :)

File: src/Data/MeasuredIn.hs (hunk 3 above)
Comment (jonathanknowles) [inline src/Data/MeasuredIn.hs]: This line is overly long, according to our coding standards. Why not wrap this as follows:

```hs
instance {-# OVERLAPS #-} (Integral i, FromJSON i) =>
    FromJSON (MeasuredIn "percent" i) where
```

File: test/unit/Cardano/Wallet/Api/TypesSpec.hs [hunk 13]
```diff
@@ -104,51 +108,50 @@ roundtripAndGolden = roundtripAndGoldenSpecsWithSettings settings
...
-    arbitrary = Amount . fromIntegral <$> (arbitrary @Word8)
+instance Arbitrary (MeasuredIn "lovelace" Natural) where
+    shrink (MeasuredIn 0) = []
+    shrink _ = [MeasuredIn 0]
+    arbitrary = MeasuredIn . fromIntegral <$> (arbitrary @Word8)
 
-instance Arbitrary Percentage where
-    arbitrary = arbitraryBoundedEnum
+instance Arbitrary (MeasuredIn "percent" Word) where
+    arbitrary = MeasuredIn <$> choose (0, 100)
```
Comment (jonathanknowles) [inline test/unit/Cardano/Wallet/Api/TypesSpec.hs]: What is wrong with using `arbitraryBoundedEnum`? We have instances of `Bounded` and `Enum` defined, so why not use them here?

File: src/Cardano/Wallet.hs (hunk 6 above)
Comment (KtorZ) [reply src/Cardano/Wallet.hs]: @Anviking @jonathanknowles 

The `Percentage` here is giving me some hard-time and there are a few points I'd like to clarify:

- The `MeasuredIn` (or `Quantity`, this is probably indeed a better name) is a polymorphic type which allows us to represent numeric quantities with units. It's not _only_ about the wiring to JSON. It happens that we do want consistency here and represent all those numeric types in the same way. It gives extra meaning to a type. Like `Key` or `Index` from the Address derivation module.

- There's nothing polymorphic about `MeasuredIn "percent" Word`. This type has only one and one concrete representation and is strictly equivalent to a `newtype Percentage = Percentage Word`. I actually went for the latter initially but it felt _wrong_ since, we already had a type to represent quantities but, we were wrapping things in a `MeasuredIn` during the serialization and deserialization. Felt weird. 

- There are actually more than one way to represent percentages. We could use a `Double` with a value between 0 and 1 for instance. So here, we really mean to represent percentage in a "human friendly" way. My point is really that, most of the types we are going to represent using `MeasuredIn` will only ever have one unit (transaction fees are going to be in `lovelace/bytes` and that's not going to change, transaction max size will be in `bytes`, and same faith). The polymorphism gives us the ability to re-use the `MeasuredIn` type to represent various more concrete types. Likewise, we could just create a newtype for each one of them, but we'll end-up duplicating the serialization logic (and testing). So, yes, we could define a `Percentage` newtype and use helper to add units to the serialization and that will do the job. But then, we do not need the `MeasuredIn` abstraction and we could simply use newtype all the time. This is just a fight for consistency here.

- The "problem" with Percentage comes from the fact that we do not have any satisfactory primitive type to represent values between 0 and 100. I agree that the overlapping instance isn't satisfactory.

With both your feedbacks, this is the two suggestions / options I see:

- We define a `Percentage` or `Percentage100` primitive type that is basically an opaque type which can hold a non-floating value going from 0 to 100. And, we do use `MeasuredIn "percent" Percentage` as a type elsewhere to makes it obvious how this type is represented for us.

- Or, we add extra `n :: Nat` parameters to the `MeasuredIn` types to contains actual bounds instead of an underlying type. So we have: `newtype MeasuredIn (min :: Nat) (max :: Nat) (unit :: Symbol)`. With this, we can define bounds directly at the type-level and represent any arbitrary bounds we would need. It becomes quite unpractical to use and we probably end-up using type-alias like `type Percentage = MeasuredIn 0 100 "percent"`  ... :/

File: src/Data/MeasuredIn.hs (hunk 11 above)
Comment (KtorZ) [reply src/Data/MeasuredIn.hs]: I'd say, if we can get stylish-haskell or hlint to do it for us, definitely. Otherwise, that will rapidly become annoying to chase those :/ 
I do like the consistency here however.

File: src/Data/MeasuredIn.hs (hunk 12 above)
Comment (KtorZ) [reply src/Data/MeasuredIn.hs]: Seems about right :sweat_smile:

File: test/unit/Cardano/Wallet/Api/TypesSpec.hs (hunk 13 above)
Comment (KtorZ) [reply test/unit/Cardano/Wallet/Api/TypesSpec.hs]: Nothing, I forgot to move the `Enum` instance back in the code. I intended to move it to the test code because that's the only place where it's actually used from at the moment. `Enum` instances are quite a pain in the *** to test, so I'd rather not give extra power to the type in our code if we don't need it. It makes it easier to define the arbitrary instance though, I totally agree.

File: src/Cardano/Wallet.hs (hunk 9 above)
Comment (KtorZ) [reply src/Cardano/Wallet.hs:322]: Not yet. It's still unclear what the `PoolId` format will end up being unfortunately. So for now, I'll keep that one "flexible" until we know better.

File: src/Cardano/Wallet.hs (hunk 5 above)
Comment (KtorZ) [reply src/Cardano/Wallet.hs:282]: Good catch :+1:

File: src/Data/MeasuredIn.hs (hunk 3 above)
Comment (KtorZ) [reply src/Data/MeasuredIn.hs]: cf answer here: https://github.com/input-output-hk/cardano-wallet/pull/107#discussion_r268066232

File: src/Cardano/Wallet.hs (hunk 6 above)
Comment (KtorZ) [reply src/Cardano/Wallet.hs]: I went for the first approach.

Comment (Anviking) [reply src/Cardano/Wallet.hs]: > but we'll end-up duplicating the serialization logic (and testing)

`DerivingVia`?

**Review by Anviking — APPROVED** [review]
Still not sure about
`!(MeasuredIn "percent" Word)`
but happy to merge!

**Comment by KtorZ** [comment]
> Still not sure about !(MeasuredIn "percent" Word) but happy to merge!

And you're write, I've fixed it locally (wrong merge conflict resolution from rebase), haven't yet pushed.
